File: Domain/ValueObjects/OrderItem/OrderItemStatus.cs
using OneOf;

namespace Domain.ValueObjects.OrderItem;

public class OrderItemStatus : ValueObject
{
    public static OrderItemStatus Pending => new OrderItemStatus("Pending");
    public static OrderItemStatus Finished => new OrderItemStatus("Finished");
    public static readonly List<OrderItemStatus> ValidStatuses = [Pending, Finished];

    public string Name { get; }

    public OrderItemStatus(string name)
    {
        Name = name;
    }

    public static OneOf<OrderItemStatus, string> CanCreate(string value)
    {
        var status = ValidStatuses.Find(status => status.Name == value);

        if (status is null)
        {
            return $"{value} is not a valid OrderItemStatus";
        }

        return status;
    }

    public static OrderItemStatus ExecuteCreate(string value)
    {
        var canCreateResult = CanCreate(value);

        if (canCreateResult.TryPickT1(out var error, out var status))
        {
            throw new Exception(error);
        }

        return status;
    }

    public static bool IsValid(string status)
    {
        return ValidStatuses.Exists(d => d.Name == status);
    }


    public override IEnumerable<object> GetEqualityComponents()
    {
        yield return Name;
    }
}

File: frontend/src/domain/valueObjects/OrderItem/OrderItemStatus.ts
export default class OrderItemStatus {
    public static readonly PENDING = new OrderItemStatus('Pending');
    public static readonly FINISHED = new OrderItemStatus('Finished');

    private constructor(value: string) {
        this.value = value;
    }

    public static create(value: string) {
        if (value === OrderItemStatus.FINISHED.value) {
            return OrderItemStatus.FINISHED;
        } else if (value === OrderItemStatus.PENDING.value) {
            return OrderItemStatus.PENDING;
        } else {
            throw new Error(`${value} is not a valid OrderItemStatus`);
        }
    }

    public value: string;
}